src/validation/rules/SingleFieldSubscriptionsRule.js

```diff
@@ -1,28 +1,119 @@
+import type {
+  OperationDefinitionNode,
+  SelectionSetNode,
+} from '../../language/ast';
+import { Kind } from '../../language/kinds';
 
 import type { ASTValidationContext } from '../ValidationContext';
 
 /**
- * Subscriptions must only include one field.
+ * Walks the selection set and returns a list of selections where extra fields
+ * were selected, and selections where introspection fields were selected.
+ */
+function walkSubscriptionSelectionSet(
+  context: ASTValidationContext,
+  selectionSet: SelectionSetNode,
+  responseKeys,
+  fieldNames = new Set(),
+  visitedFragmentNames = new Set(),
+  extraFieldSelections = [],
+  introspectionFieldSelections = [],
+) {
+  for (const selection of selectionSet.selections) {
+    switch (selection.kind) {
+      case Kind.FIELD: {
+        const fieldName = selection.name.value;
+        fieldNames.add(fieldName);
+        const responseName = selection.alias
+          ? selection.alias.value
+          : fieldName;
+        responseKeys.add(responseName);
+        if (fieldName[0] === '_' && fieldName[1] === '_') {
+          // fieldName represents an introspection field if it starts with `__`
+          introspectionFieldSelections.push(selection);
+        } else if (fieldNames.size > 1 || responseKeys.size > 1) {
+          extraFieldSelections.push(selection);
+        }
+        break;
+      }
+      case Kind.FRAGMENT_SPREAD: {
+        const fragmentName = selection.name.value;
+        if (!visitedFragmentNames.has(fragmentName)) {
+          visitedFragmentNames.add(fragmentName);
+          const fragment = context.getFragment(fragmentName);
+          if (fragment) {
+            walkSubscriptionSelectionSet(
+              context,
+              fragment.selectionSet,
+              responseKeys,
+              fieldNames,
+              visitedFragmentNames,
+              extraFieldSelections,
+              introspectionFieldSelections,
+            );
+          }
+        }
+        break;
+      }
+      case Kind.INLINE_FRAGMENT: {
+        walkSubscriptionSelectionSet(
+          context,
+          selection.selectionSet,
+          responseKeys,
+          fieldNames,
+          visitedFragmentNames,
+          extraFieldSelections,
+          introspectionFieldSelections,
+        );
+        break;
+      }
+    }
+  }
+  return [extraFieldSelections, introspectionFieldSelections];
+}
+
+/**
+ * Subscriptions must only include a non-introspection field.
  *
- * A GraphQL subscription is valid only if it contains a single root field.
+ * A GraphQL subscription is valid only if it contains a single root field and
+ * that root field is not an introspection field.
  */
 export function SingleFieldSubscriptionsRule(
   context: ASTValidationContext,
 ): ASTVisitor {
   return {
     OperationDefinition(node: OperationDefinitionNode) {
       if (node.operation === 'subscription') {
-        if (node.selectionSet.selections.length !== 1) {
```

Comment:
Happy to also see this validation bug being fixed - fields are collected before asserting a single selection